src/app/clusters/scenes-server/scenes-server.cpp

```diff
@@ -96,6 +98,202 @@ CHIP_ERROR UpdateLastConfiguredBy(HandlerContext & ctx, ResponseType resp)
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Clears the SceneInfoStruct associated to a fabric and compresses the array to leave uninitialised structs at the end
+/// @param[in] endpoint target endpoint
+/// @param[in] fabric target fabric
+void ScenesServer::FabricSceneInfo::ClearSceneInfoStruct(EndpointId endpoint, FabricIndex fabric)
+{
+    size_t endpointIndex = 0;
+    ReturnOnFailure(FindFabricSceneInfoIndex(endpoint, endpointIndex));
+    uint8_t sceneInfoStructIndex = 0;
+    ReturnOnFailure(FindSceneInfoStructIndex(fabric, endpointIndex, sceneInfoStructIndex));
+
+    uint8_t nextIndex = static_cast<uint8_t>(sceneInfoStructIndex + 1);
+    uint8_t moveNum   = static_cast<uint8_t>(ArraySize(mSceneInfoStructs[endpointIndex]) - nextIndex);
+    // Compress the endpoint's SceneInfoStruct array
+    if (moveNum)
+    {
+        for (size_t i = 0; i < moveNum; ++i)
+        {
+            mSceneInfoStructs[endpointIndex][sceneInfoStructIndex + i] = mSceneInfoStructs[endpointIndex][nextIndex + i];
+        }
+    }
+
+    // Decrement the SceneInfoStruct count
+    mSceneInfoStructsCount[endpointIndex]--;
+
+    // Clear the last populated SceneInfoStruct
+    mSceneInfoStructs[endpointIndex][mSceneInfoStructsCount[endpointIndex]].fabricIndex       = kUndefinedFabricIndex;
+    mSceneInfoStructs[endpointIndex][mSceneInfoStructsCount[endpointIndex]].sceneCount        = 0;
+    mSceneInfoStructs[endpointIndex][mSceneInfoStructsCount[endpointIndex]].currentScene      = 0;
+    mSceneInfoStructs[endpointIndex][mSceneInfoStructsCount[endpointIndex]].currentGroup      = 0;
+    mSceneInfoStructs[endpointIndex][mSceneInfoStructsCount[endpointIndex]].remainingCapacity = 0;
+}
+
+/// @brief Returns the index of the FabricSceneInfo associated to an endpoint
+/// @param[in] endpoint target endpoint
+/// @param[out] endpointIndex index of the corresponding FabricSceneInfo for an endpoint, corresponds to a row in the
+/// mSceneInfoStructs array,
+/// @return CHIP_NO_ERROR or CHIP_ERROR_NOT_FOUND, CHIP_ERROR_INVALID_ARGUMENT if invalid endpoint
+CHIP_ERROR ScenesServer::FabricSceneInfo::FindFabricSceneInfoIndex(EndpointId endpoint, size_t & endpointIndex)
+{
+    VerifyOrReturnError(kInvalidEndpointId != endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+
+    uint16_t index = emberAfGetClusterServerEndpointIndex(endpoint, Scenes::Id, EMBER_AF_SCENES_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (index < ArraySize(mSceneInfoStructs))
+    {
+        endpointIndex = index;
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+/// @brief Returns the SceneInfoStruct associated to a fabric
+/// @param[in] fabric target fabric index
+/// @param[in] endpointIndex index of the corresponding FabricSceneInfo for an endpoint, corresponds to a row in the
+/// mSceneInfoStructs array
+/// @param[out] index index of the corresponding SceneInfoStruct if found, mFabricSceneInfo[endpoint] out of bounds index otherwise
```

Comment:
How about just saying that if not found the value of this outparam should not be assumed to be anything in particular and callers must not use it?  Or do they in fact use it?